server.test.js

```diff
@@ -0,0 +1,135 @@
+const url = require('url')
+const { processRedirectString, getRedirect } = require('./server')
+
+describe('processRedirectString', () => {
+  it('reads the regex, replacement and code', () => {
+    const { regex, replace, code } = processRedirectString('^/foo /bar 418')
+
+    expect(regex).toBeInstanceOf(RegExp)
+    expect(regex.source).toEqual('^\\/foo')
+    expect(replace).toEqual('/bar')
+    expect(code).toEqual(418)
+  })
+  it('code defaults to 301', () => {
+    const { code } = processRedirectString('^/x /x')
+
+    expect(code).toEqual(301)
+  })
+  it('detects whether we are matching a pathname or the whole url', () => {
+    const { matchPathname } = processRedirectString('^/pathname /x')
+    expect(matchPathname).toEqual(true)
+
+    const { matchPathname: matchPathnameFalse } = processRedirectString(
+      '^https://example.com/foo /x'
+    )
+    expect(matchPathnameFalse).toEqual(false)
+  })
+})
+
+describe('getRedirect', () => {
+  it('redirects to https while removing www', () => {
+    const fakeReq = {
+      headers: {
+        'x-forwarded-proto': 'http'
+      },
+      url: '/foo/bar?baz'
+    }
+    expect(
+      getRedirect('www.example.com', '/not-used', { req: fakeReq, dev: false })
+    ).toEqual([301, 'https://example.com/foo/bar?baz'])
+  })
+  const itRedirects = (source, target, code = 301) => {
+    const addHost = pathOrUrl => {
+      if (pathOrUrl.startsWith('/')) {
+        return `https://dvc.org${pathOrUrl}`
+      }
+      return pathOrUrl
+    }
+
+    it(`${source} -> ${target} (${code})`, () => {
+      source = addHost(source)
+      const { hostname, pathname } = url.parse(source)
+      const [rCode, rLocation] = getRedirect(hostname, pathname)
+
+      expect(rLocation).toEqual(target)
+      expect(rCode).toEqual(code)
+
+      // Detect redirect loops
+      const secondUrl = url.parse(addHost(rLocation))
+      const secondRedirect = getRedirect(secondUrl.hostname, secondUrl.pathname)
+      expect(secondRedirect).toEqual([])
+    })
+  }
+
+  describe('host redirects', () => {
```

Comment:
@jorgeorpinel could you take a second look please, that we got all the redirects right